File: Solution/0236.Lowest_Common_Ancestor_of_a_Binary_Tree/0236.Lowest_Common_Ancestor_of_a_Binary_Tree.py
#递归法
class Solution:
    def __init__(self):
        self.ans=None
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        self.dfs(root,p,q)
        return self.ans
    
    def dfs(self,root,p,q):
        if root is None:
            return False
        flson=self.dfs(root.left,p,q)
        frson=self.dfs(root.right,p,q)
        if (flson and frson) or ((root.val==p.val or root.val==q.val) and (flson or frson)):
            self.ans=root
        return flson or frson or root==p or root==q
      
 #存储父节点
class Solution:
    def __init__(self):
        self.fa=dict()
        self.vis=dict()
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        self.fa[root.val]=None
        self.dfs(root)
        while p:
            self.vis[p.val]=True
            p=self.fa[p.val]
        while q:
            if q.val in self.vis:
                return q
            else:
                q=self.fa[q.val]
        return None
    def dfs(self,root):
        if root.left:
            self.fa[root.left.val]=root
            self.dfs(root.left)
        if root.right:
            self.fa[root.right.val]=root
            self.dfs(root.right)

File: Solution/0236.Lowest_Common_Ancestor_of_a_Binary_Tree/0236.Lowest_Common_Ancestor_of_a_Binary_Tree.cpp
//法一:递归法
class Solution {
public:
    TreeNode *ans;
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        dfs(root,p,q);
        return ans;
    }
    bool dfs(TreeNode* root,TreeNode* p,TreeNode* q)
    {
        if(!root)
            return false;
        bool flson=dfs(root->left,p,q);
        bool frson=dfs(root->right,p,q);
        if((flson&&frson)||((root==p||root==q)&&(flson||frson)))
            ans=root;
        return root==p||root==q||flson||frson;
    }
};

//存储父节点
class Solution {
public:
    unordered_map<int,TreeNode*> fa;
    unordered_map<int,bool> vis;
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        fa[root->val]=NULL;
        dfs(root);
        while(p)
        {
            vis[p->val]=true;
            p=fa[p->val];
        }
        while(!vis[q->val])
        {
            q=fa[q->val];
        }
        return q;
    }
    void dfs(TreeNode* root)
    {
        if(root->left)
        {
            fa[root->left->val]=root;
            dfs(root->left);
        }
        if(root->right)
        {
            fa[root->right->val]=root;
            dfs(root->right);
        }
    }
};
